pkg/controller/constrainttemplate/constrainttemplate_controller.go

```diff
@@ -453,8 +494,100 @@ func (r *ReconcileConstraintTemplate) handleUpdate(
+		if err := r.Get(ctx, types.NamespacedName{Name: vapName}, currentVap); err != nil {
+			if !apierrors.IsNotFound(err) && !errors.As(err, &discoveryErr) && !meta.IsNoMatchError(err) {
+				return reconcile.Result{}, err
+			}
+			currentVap = nil
+		}
+		logger.Info("get vap", "vapName", vapName, "currentVap", currentVap)
+		newVap := &admissionregistrationv1beta1.ValidatingAdmissionPolicy{}
+		transformedVap, err := transform.TemplateToPolicyDefinition(unversionedCT)
+		if err != nil {
+			logger.Info("transform to vap error", "vapName", vapName, "error", err)
+			createErr := &v1beta1.CreateCRDError{Code: ErrCreateCode, Message: err.Error()}
+			status.Status.Errors = append(status.Status.Errors, createErr)
+			err := r.reportErrorOnCTStatus(ctx, ErrCreateCode, "Could not transform to vap object", status, err)
+			return reconcile.Result{}, err
+		}
+		if currentVap == nil {
+			newVap = transformedVap.DeepCopy()
+		} else {
+			newVap = currentVap.DeepCopy()
+			newVap.Spec = transformedVap.Spec
+		}
+
+		if err := controllerutil.SetControllerReference(ct, newVap, r.scheme); err != nil {
+			return reconcile.Result{}, err
+		}
+
+		if currentVap == nil {
+			logger.Info("creating vap", "vapName", vapName)
+			if err := r.Create(ctx, newVap); err != nil {
+				logger.Info("creating vap error", "vapName", vapName, "error", err)
+				createErr := &v1beta1.CreateCRDError{Code: ErrCreateCode, Message: err.Error()}
+				status.Status.Errors = append(status.Status.Errors, createErr)
+				err := r.reportErrorOnCTStatus(ctx, ErrCreateCode, "Could not create vap object", status, err)
+				return reconcile.Result{}, err
+			}
+		}
+		if !reflect.DeepEqual(currentVap, newVap) {
+			logger.Info("updating vap")
+			if err := r.Update(ctx, newVap); err != nil {
+				updateErr := &v1beta1.CreateCRDError{Code: ErrUpdateCode, Message: err.Error()}
+				status.Status.Errors = append(status.Status.Errors, updateErr)
+				err := r.reportErrorOnCTStatus(ctx, ErrUpdateCode, "Could not update vap object", status, err)
+				return reconcile.Result{}, err
+			}
+
+			// after vap is updated, trigger update event for all constraints
+			gvk := makeGvk(ct.Spec.CRD.Spec.Names.Kind)
+			logger.Info("list gvk objects", "gvk", gvk)
+			cstrObjs, err := r.listObjects(ctx, gvk)
+			if err != nil {
+				logger.Error(err, "get all constraints listObjects")
+				updateErr := &v1beta1.CreateCRDError{Code: ErrUpdateCode, Message: err.Error()}
+				status.Status.Errors = append(status.Status.Errors, updateErr)
+				err := r.reportErrorOnCTStatus(ctx, ErrUpdateCode, "Could not list all constraint objects", status, err)
+				return reconcile.Result{}, err
+			}
+			logger.Info("list gvk objects", "cstrObjs", cstrObjs)
+			for _, cstr := range cstrObjs {
+				c := cstr
+				logger.Info("triggering cstrEvent")
+				r.cstrEvents <- event.GenericEvent{Object: &c}
+			}
+		}
+	}
+	// do not generate vap resources
+	// remove if exists
+	if !generateVap && constraint.IsVapAPIEnabled() {
```

Comment:
We should add the "listObjects/replay constraints" behavior here as well (might make sense to factor that out into its own function to avoid code duplication)

Reply:
Thinking more deeply about this, I think we have a general race condition. Let's pretend:

1) constraint controller is fast
2) CT controller is slow (likely, since it may require compilation of Rego/CEL)
3) there is some constraint Foo
4) It is backed by some template FooTemplate
5) there are 3 events for FooTemplate: no-generate-vap, generate-vap, no-generate-vap
6) FooTemplate reconciler is slow enough that only two events get reconciled: no-generate-vap, no-generate-vap
7) constraint controller gets a reconcile while generate-vap is set

If this sequence of events happens, you could get a dangling VAP binding, b/c when constraint controller reconciles, it's gonna run `client.Get()` on the informer cache and see the annotation to create a binding, but the constraint template controller sees no change (from it's view `no-generate-vap` has always been set), so it will not trigger a reconcile.

I think I see a way out of this, but it's not super pretty:

1) create a shared cache between the constraint template and constraint controllers that stores the CT controller's opinion as to whether a constraint kind should/should not generate VAP objects
2) create a function on the CT reconciler called `shouldGenerateVAP(kind string, shouldGenerate bool)`, when `shouldGenerate` changes for a particular kind, that's what triggers a re-reconcile of the constraint
3) call `shouldGenerateVAP()` immediately after determining `generateVap`
4) Because the results of reconciliation are dependent on internal state, we should run reconciliation as a singleton process (i.e. throw it on the status/audit Pod), this will prevent thrashing of the k8s resources whenever the pods' watch events are out of sync

(4) is probably a good thing to address generally, since there is also a chance of thrashing on G8r upgrade. I'd suggest that any controller that creates a downstream object (CRDs, VAP objects), the creation runs as a singleton (we can carve it out as an operation, pehaps, like audit or status updates). The controllers should still run for the purposes of caching state (e.g. loading constraint templates into CF), and they should still write out pod status objects.

This is medium-to-large body of work. I would recommend we do this before graduating to beta (or at least before allowing users to set failure policy to Fail). However, I don't think we need to block the current PR on this because:

* This is an in-development feature that is off-by-default
* We currently fail open, so a dangling binding will not cause any issues